libs/@local/harpc/server/src/router.rs

```diff
@@ -0,0 +1,244 @@
+use alloc::sync::Arc;
+use core::{
+    future::{self, Ready},
+    task::{Context, Poll},
+};
+
+use frunk::{HCons, HNil};
+use futures::future::Either;
+use harpc_net::codec::ErrorEncoder;
+use harpc_service::delegate::ServiceDelegate;
+use harpc_tower::{
+    body::{Body, BodyExt},
+    request::Request,
+    response::{BoxedResponse, Parts, Response},
+};
+use harpc_types::{service::ServiceId, version::Version};
+use tower::{Layer, Service, ServiceBuilder, ServiceExt, layer::util::Identity, util::Oneshot};
+
+use crate::{delegate::ServiceDelegateHandler, error::NotFound, session::SessionStorage};
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub struct Handler<S> {
+    service: ServiceId,
+    version: Version,
+
+    inner: S,
+}
+
+pub trait Route<B, C> {
+    type ResponseBodyError;
+    type Future: Future<Output = Result<BoxedResponse<Self::ResponseBodyError>, !>> + Send;
```

Comment:
Why do we require the `Future` to have a `!` in the `Err` variant? Can't we just use the `Ok` part instead?

Reply:
see: [1e966ef](https://github.com/hashintel/hash/pull/5320/commits/1e966efcdacabb7210df7f00d2f6e94f25168973)